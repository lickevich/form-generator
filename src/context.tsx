import React, { useState } from "react";

import defaultConfig from "./data";
import { ContextType } from "./type";
import { ROUTES } from "./constants";
import { isIncludedValue } from "./utils/helpers";

export const initialState = {
  template: {
    title: "",
    items: [],
    buttons: [],
  },
  setTemplate: () => {},
  error: "",
  setError: () => {},
  config: defaultConfig,
  setConfig: () => {},
  blockedRoutesIds: [ROUTES.RESULT.ID],
  setBlockedRoutesIds: () => {},
  isResultAvailable: false,
};

export const Context = React.createContext<ContextType>(initialState);

export const ContextProvider: React.FC = ({ children }) => {
  const [template, setTemplate] = useState<ContextType["template"]>(
    initialState.template
  );
  const [error, setError] = useState<ContextType["error"]>(initialState.error);
  const [config, setConfig] = useState<ContextType["config"]>(
    initialState.config
  );
  const [blockedRoutesIds, setBlockedRoutesIds] = useState<
    ContextType["blockedRoutesIds"]
  >(initialState.blockedRoutesIds);
  const isResultAvailable: boolean = !isIncludedValue(
    blockedRoutesIds,
    ROUTES.RESULT.ID
  );

  return (
    <Context.Provider
      value={{
        template,
        setTemplate,
        error,
        setError,
        config,
        setConfig,
        blockedRoutesIds,
        setBlockedRoutesIds,
        isResultAvailable,
      }}
    >
      {children}
    </Context.Provider>
  );
};

export default ContextProvider;
