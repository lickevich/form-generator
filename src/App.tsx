import React, { useContext } from "react";
import { Switch, Route, Redirect } from "react-router-dom";

import { Context } from "./context";
import { ContextType } from "./type";
import routes from "./routes";
import { Container, Main } from "./style";
import Tabs from "./components/Tabs";
import Notification from "./components/Notification";
import Config from "./containers/Config";
import Result from "./containers/Result";
import { ROUTES } from "./constants";

const App: React.FC = () => {
  const { error, isResultAvailable } = useContext(Context) as ContextType;

  return (
    <Container>
      <Notification message={error} />
      <Tabs routes={routes} />
      <Main>
        <Switch>
          <Route exact path={ROUTES.CONFIG.PATH}>
            <Config />
          </Route>
          <Route exact path={ROUTES.RESULT.PATH}>
            {isResultAvailable ? (
              <Result />
            ) : (
              <Redirect to={ROUTES.CONFIG.PATH} />
            )}
          </Route>
        </Switch>
      </Main>
    </Container>
  );
};

export default App;
