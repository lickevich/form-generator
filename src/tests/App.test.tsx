import React from "react";
import { render, cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

import { ContextProvider } from "../context";
import App from "../App";

const renderWithContext = (component: JSX.Element) => {
  return {
    ...render(<ContextProvider>{component}</ContextProvider>),
  };
};

afterEach(cleanup);

describe("App", () => {
  it("render the component", () => {
    const component = renderWithContext(
      <Router>
        <App />
      </Router>
    );
    const { asFragment } = component;

    // @ts-ignoreq
    expect(asFragment(component)).toMatchSnapshot();
  });
});
