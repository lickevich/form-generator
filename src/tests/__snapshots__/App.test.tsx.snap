// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render the component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvaa dLwSED"
  >
    <div
      class="sc-hKwCoD enIAsM"
    />
    <a
      aria-current="page"
      class="sc-dkPtyc izQPWx selected"
      href="/"
    >
      Config
    </a>
    <a
      class="sc-dkPtyc izQPWx"
      href="/result"
    >
      Result
    </a>
    <div
      class="sc-gsDJrp FIKdM"
    >
      <form>
        <div
          class="sc-gKckTs czEeVz"
        >
          <textarea
            class="sc-eCImvq ciqZQD"
          >
            {
  "title": "Form Title",
  "items": [
    { "type": "text", "label": "textfield" },
    { "type": "number", "label": "numberfield" },
    { "type": "textarea", "label": "textarea" },
    { "type": "date", "label": "datefield" },
    {
      "type": "checkbox",
      "groupLabel": "checkbox",
      "items": [{ "label": "checkbox1" }, { "label": "checkbox2" }]
    },
    {
      "type": "radio",
      "groupLabel": "radio",
      "items": [
        { "label": "radio1", "name": "radio", "value": "1" },
        { "label": "radio2", "name": "radio", "value": "2" }
      ]
    }
  ],
  "buttons": [
    { "type": "submit", "value": "Apply" },
    { "type": "button", "value": "Ok" },
    { "type": "reset", "value": "Reset" }
  ]
}
          </textarea>
        </div>
        <div
          class="sc-iCfLBT imxnEl"
        >
          <input
            class="sc-jRQAMF EJfAC"
            id=""
            type="submit"
            value="Apply"
          />
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
